test/bdd/pkg/edv/edv_steps.go

```diff
@@ -49,71 +73,325 @@ func (e *Steps) createDataVault(vaultID, expectedVaultLocation string) error {
+
+	retrievedDocumentBytes, err := client.ReadDocument(vaultID, docID)
+	if err != nil {
+		return err
 	}
 
-	docLocation, err := client.CreateDocument(vaultID, &document)
+	retrievedDocument := operation.EncryptedDocument{}
+
+	err = json.Unmarshal(retrievedDocumentBytes, &retrievedDocument)
 	if err != nil {
 		return err
 	}
 
-	if docLocation != expectedDocLocation {
-		return fmt.Errorf("expected the new data vault location to be %s, got %s instead", expectedDocLocation, docLocation)
+	err = verifyEncryptedDocsAreEqual(retrievedDocument, e.bddContext.EncryptedDocToStore)
+	if err != nil {
+		return err
 	}
 
+	e.bddContext.ReceivedEncryptedDoc = retrievedDocument
+
 	return nil
 }
 
-func (e *Steps) vaultHasDocument(vaultID, docID string) error {
-	err := e.createDocument(docID, vaultID, "localhost:8080/encrypted-data-vaults/"+vaultID+"/docs/"+docID)
+func (e *Steps) decryptDocument() error {
+	decryptedEnvelope, err := e.bddContext.Packer.Unpack(e.bddContext.ReceivedEncryptedDoc.JWE)
+	if err != nil {
+		return err
+	}
+
+	decryptedDoc := operation.StructuredDocument{}
+
+	err = json.Unmarshal(decryptedEnvelope.Message, &decryptedDoc)
+	if err != nil {
+		return err
+	}
+
+	err = verifyStructuredDocsAreEqual(decryptedDoc, e.bddContext.StructuredDocToBeEncrypted)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (e *Steps) receiveDocument(docID, vaultID, expectedDocumentArgKey string) error {
-	client := edv.New(e.bddContext.EDVHostURL)
+func verifyEncryptedDocsAreEqual(retrievedDocument, expectedDocument operation.EncryptedDocument) error {
+	if retrievedDocument.ID != expectedDocument.ID {
+		return unexpectedValueError(expectedDocument.ID, retrievedDocument.ID)
+	}
+
+	if retrievedDocument.Sequence != expectedDocument.Sequence {
+		return unexpectedValueError(string(expectedDocument.Sequence), string(retrievedDocument.Sequence))
+	}
 
-	document, err := client.ReadDocument(vaultID, docID)
+	err := verifyJWEFieldsAreEqual(expectedDocument, retrievedDocument)
 	if err != nil {
 		return err
 	}
 
-	expectedDocument := e.bddContext.Args[expectedDocumentArgKey]
+	return nil
+}
 
-	if string(document) != expectedDocument {
-		return fmt.Errorf("expected the document to be %s, got %s instead", expectedDocument, string(document))
+func verifyJWEFieldsAreEqual(expectedDocument, retrievedDocument operation.EncryptedDocument) error {
+	// CouchDB likes to switch around the order of the fields in the JSON.
+	// This means that we can't do a direct string comparison of the JWE (json.rawmessage) fields
+	// in the EncryptedDocument structs. Instead we need to check each field manually.
+	var expectedJWEFields map[string]string
+
+	err := json.Unmarshal(expectedDocument.JWE, &expectedJWEFields)
+	if err != nil {
+		return err
+	}
+
+	expectedProtectedFieldValue, expectedIVFieldValue, expectedCiphertextFieldValue, expectedTagFieldValue,
+		err := getJWEFieldValues(expectedJWEFields, "expected JWE")
```

Comment:
maybe check if there exists a JSON Compare() function in a library instead of comparing manually the fields this way.

Reply:
Unfortunately the built in JSON library doesn't have this type of functionality.

After doing some searching, I've come up with a few alternatives...

**Idea 1:**

I could unmarshal the JSON to a map[string]interface{}, and then use reflect.DeepEqual() on them https://golang.org/pkg/reflect/#DeepEqual.

**Idea 2:**

I could unmarshal the JSON to a map[string]interface{}, and then marshal back to an []byte. Then I can string compare them, because according to the documentation they should be in the same order.

Do you think either one of these ideas is better than what I've originally written? I'm leaning towards either Idea 2 or keeping the code as is, but I'm not sure. @Baha-sk @sandrask

Reply:
keep code as is then.. DeepEqual is recursive and doing Unmarshal JSON into map then marshal back to a []byte means is not a clean way to check for equality.